### PR TITLE
[RFC] hotplug: handling of simple add/remove scenario

### PR DESCRIPTION
**Pull request by stolowski**
This is the integration branch for hotplug with RFC. I'll be splitting it into smaller PRs (some of that is already proposed) and mark this one blocked.
Note: the udevmon code here will most likely be moved into separate package, this is being discussed and happening in PR #5618 and will be reflected here once fully ready.

The feature is locked by `experimental.hotplug` feature flag. For the time being it implements the simple "happy" scenario, that is: hotplug 'add' event gets received (either from hotplug event, or during device enumeration on startup) and respective slot(s) are created (at the moment there is just one "hotplug" aware interface in this branch - serial-port) and available for the user; if the device was seen before (had connections), the connections will be recreated. 'remove' hotplug event is handled as long as the device is unplugged when snapd is running, in which case slots get disconnected and removed from the repository.

What's missing:
- tests (WIP).
- an algorithm for generating nice names for the slots.
- handling of autoconnect slots
- finding about removed devices during startup enumeration, and feeding them to remove handler.
- edge cases (unclear yet).

### CODE REVIEW

File: interfaces/repo.go [hunk 1]
```diff
@@ -763,12 +782,49 @@ func (r *Repository) connected(snapName, plugOrSlotName string) ([]*ConnRef, err
 	return conns, nil
 }
 
+// ConnectionsForDeviceKey returns all hotplug connections for given device key and interface name.
+func (r *Repository) ConnectionsForDeviceKey(deviceKey, ifaceName string) ([]*ConnRef, error) {
+	snapName, err := r.GuessSystemSnapName()
```
Comment (zyga) [inline interfaces/repo.go]: Please lock the repository mutex here!

File: overlord/ifacestate/handlers.go [hunk 2]
```diff
@@ -1118,3 +1119,172 @@ func (m *InterfaceManager) doGadgetConnect(task *state.Task, _ *tomb.Tomb) error
...
+
+// doHotplugConnect creates task(s) to (re)create connections in response to hotplug "add" event.
+func (m *InterfaceManager) doHotplugConnect(task *state.Task, _ *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	conns, err := getConns(st)
+	if err != nil {
+		return fmt.Errorf("failed to read connections data: %s", err)
```
Comment (zyga) [inline overlord/ifacestate/handlers.go]: I think we should consistently use `cannot ...` pattern for error messages. How about `cannot retrieve connections`. Alternatively we could just return the error as-is. Was there a specific reason to prefix the returned error?

File: interfaces/builtin/serial_port.go [hunk 3]
```diff
@@ -208,6 +209,25 @@ func (iface *serialPortInterface) AutoConnect(*snap.PlugInfo, *snap.SlotInfo) bo
 	return true
 }
 
+func (iface *serialPortInterface) HotplugDeviceDetected(di *hotplug.HotplugDeviceInfo, spec *hotplug.Specification) error {
+	if di.Subsystem() == "tty" && strings.HasPrefix(di.DeviceName(), "/dev/ttyUSB") {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply interfaces/builtin/serial_port.go]: Yes, we can do this. Note, at the moment I added a simple top-level filtering which reduces the noise received from go-udev and udevadm - look for `filter = &netlink.RuleDefinitions{` below.

File: interfaces/builtin/serial_port.go [hunk 4]
```diff
@@ -208,6 +209,25 @@ func (iface *serialPortInterface) AutoConnect(*snap.PlugInfo, *snap.SlotInfo) bo
...
+		seqnum, ok := di.Attribute("SEQNUM")
+		if !ok {
+			// FIXME
+			seqnum = "0"
```
[the remark this answers is not in the document]
Comment (stolowski) [reply interfaces/builtin/serial_port.go]: Right, we probably should.
FWTW this bit here is about inventing slot names and as indicated in the TODO of this PR, I intend to implement a better helper method for generating them (e.g. derive them from product names and other readable strings, after cleanup).

File: overlord/ifacestate/handlers.go (hunk 2 above)
Comment (stolowski) [reply overlord/ifacestate/handlers.go]: Right the prefix is not really needed here, I missed the fact that getConns already returns a meaningful error message. Removed.

File: interfaces/builtin/serial_port.go (hunk 3 above)
Comment (bboozzoo) [inline interfaces/builtin/serial_port.go]: I think this may be more than just `/dev/ttyUSB` interfaces, there's a whole zoo of `tty<somerandomstring>` appearing on various SoC systems.

File: overlord/ifacestate/hotplug.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+			logger.Debugf("No valid device key provided by interface %s, device %q ignored", iface.Name(), devinfo.DeviceName())
+			continue
+		}
+
+		if !hotplugFeature {
+			logger.Noticef("Hotplug 'add' event for device %q (interface %q) ignored, enable experimental.hotplug", devinfo.DevicePath(), iface.Name())
+			continue
+		}
+
+		logger.Debugf("HotplugDeviceAdded: %s (default key: %q, devname %s, subsystem: %s)", devinfo.DevicePath(), defaultKey, devinfo.DeviceName(), devinfo.Subsystem())
```
Comment (bboozzoo) [inline overlord/ifacestate/hotplug.go]: Since it's per interface, maybe adding some iface name would be useful (just in case they have the same keys or end up using default).

File: overlord/ifacestate/hotplug.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+		if len(slotSpecs) == 0 {
+			continue
+		}
+
+		if !validateDeviceKey(key) {
+			logger.Debugf("No valid device key provided by interface %s, device %q ignored", iface.Name(), devinfo.DeviceName())
+			continue
+		}
+
+		if !hotplugFeature {
```
Comment (bboozzoo) [inline overlord/ifacestate/hotplug.go]: This might go before the loop, if it's disabled then it's disabled for every interface.

File: overlord/ifacestate/hotplug.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+			}
+
+			if err := m.repo.AddSlot(slot); err != nil {
+				logger.Noticef("Failed to create slot %q for interface %s", slot.Name, slot.Interface)
+				continue
+			}
+			logger.Noticef("Added hotplug slot %q of interface %s for device key %q", slot.Name, slot.Interface, key)
+		}
+
+		chg := st.NewChange(fmt.Sprintf("hotplug-connect-%s", iface), fmt.Sprintf("Connect hotplug slots of interface %s", iface))
```
Comment (bboozzoo) [inline overlord/ifacestate/hotplug.go]: Won't this be happening every time snapd starts?

File: overlord/ifacestate/hotplug.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+		}
+		connsForDevice = append(connsForDevice, id)
+	}
+	return connsForDevice
+}
+
+func (m *InterfaceManager) hotplugEnabled() (bool, error) {
+	tr := config.NewTransaction(m.state)
+	var featureFlagHotplug bool
+	if err := tr.GetMaybe("core", "experimental.hotplug", &featureFlagHotplug); err != nil {
```
Comment (bboozzoo) [inline overlord/ifacestate/hotplug.go]: `snapstate` has a similar helper `getFeatureFlagBool` that handles a case when the feature was disabled via `snap set system experimenta.feature=` in which case deserialization to `bool` will fail. Maybe you could reuse that code or make it public.

File: overlord/ifacestate/hotplug.go (hunk 6 above)
Comment (stolowski) [reply overlord/ifacestate/hotplug.go]: It's deliberately done that way, the goal is to excercise as much of the hotplug logic as possible even if the feature is disabled - i.e. avoid any modifications of the state or interfaces, but otherwise execute the code. If we wanted to cut the feature off completely based on the feature flag, this would be more effective on a higher level.

File: interfaces/builtin/serial_port.go (hunk 3 above)
Comment (stolowski) [reply interfaces/builtin/serial_port.go]: Yes, I'm aware of that, we use "^/dev/tty(mxc|USB|ACM|AMA|XRUSB|S|O)[0-9]+$" pattern for the non-hotplug-aware part of this interface. I just went for a simple ttyUSB" as this is the only one I can reasonably test at the moment.
Anyway, noted, thanks, I'll either extend this with a regex too, or remove completely if we don't want to only partially support hotplug for this interface.

File: overlord/ifacestate/hotplug.go (hunk 8 above)
Comment (stolowski) [reply overlord/ifacestate/hotplug.go]: Ah indeed, I forgot we had a helper already, thanks! I'll make it public.

**Comment by stolowski** [comment]
Closing for now as this is very close to be proposed in 2-3 separate PRs.
